File: extension/lib/tvmaze.js
// TVMaze API client (free, no API key, CORS-open).

const API_BASE = "https://api.tvmaze.com";
const DEFAULT_RUNTIME = 45;

export class TVMazeError extends Error {}

export function stripHtml(text) {
  if (!text) return "";
  const noTags = text.replace(/<[^>]+>/g, "");
  // Minimal entity handling for TVMaze summaries (no DOM in service workers).
  return noTags
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#0?39;|&apos;/g, "'")
    .replace(/&nbsp;/g, " ")
    .trim();
}

async function get(path, params) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [k, v] of Object.entries(params || {})) url.searchParams.set(k, v);
  let resp;
  try {
    resp = await fetch(url);
  } catch (err) {
    throw new TVMazeError(`TVMaze request failed: ${err.message}`);
  }
  if (resp.status === 404) throw new TVMazeError("No show found matching that name.");
  if (!resp.ok) throw new TVMazeError(`TVMaze error: HTTP ${resp.status}`);
  return resp.json();
}

export async function searchShows(query) {
  const results = await get("/search/shows", { q: query });
  return results.map(({ show }) => ({
    id: show.id,
    name: show.name,
    premiered: (show.premiered || "").slice(0, 4),
    status: show.status,
    genres: show.genres || [],
  }));
}

export function normalizeShow(data) {
  const fallbackRuntime = data.averageRuntime || DEFAULT_RUNTIME;
  const rawEps = data._embedded?.episodes || [];
  const episodes = rawEps
    .filter((e) => e.season != null && e.number != null)
    .map((e) => ({
      id: e.id,
      season: e.season,
      number: e.number,
      title: e.name || `Episode ${e.number}`,
      runtime: e.runtime || fallbackRuntime,
      airdate: e.airdate || "",
      summary: stripHtml(e.summary),
      type: e.type || "regular",
    }))
    .sort((a, b) => a.season - b.season || a.number - b.number);
  return {
    id: data.id,
    name: data.name,
    premiered: (data.premiered || "").slice(0, 4),
    status: data.status,
    genres: data.genres || [],
    episodes,
  };
}

// Best-matching show with normalized, ordered episode list.
export async function getShow(query) {
  const data = await get("/singlesearch/shows", { q: query, embed: "episodes" });
  return normalizeShow(data);
}

File: src/bingebreak/tvmaze.py
"""TVMaze API client (free, no API key required)."""

from __future__ import annotations

import html
import re

import requests

API_BASE = "https://api.tvmaze.com"
DEFAULT_RUNTIME = 45  # minutes, when TVMaze has no runtime for an episode

_TAG_RE = re.compile(r"<[^>]+>")


class TVMazeError(RuntimeError):
    """Raised when TVMaze lookup fails."""


def strip_html(text: str | None) -> str:
    if not text:
        return ""
    return html.unescape(_TAG_RE.sub("", text)).strip()


def _get(path: str, params: dict | None = None) -> dict | list:
    try:
        resp = requests.get(f"{API_BASE}{path}", params=params, timeout=15)
    except requests.RequestException as exc:
        raise TVMazeError(f"TVMaze request failed: {exc}") from exc
    if resp.status_code == 404:
        raise TVMazeError("No show found matching that name.")
    resp.raise_for_status()
    return resp.json()


def search_shows(query: str) -> list[dict]:
    """Return candidate shows for a query, best match first."""
    results = _get("/search/shows", {"q": query})
    shows = []
    for item in results:
        show = item.get("show", {})
        shows.append(
            {
                "id": show.get("id"),
                "name": show.get("name"),
                "premiered": (show.get("premiered") or "")[:4],
                "status": show.get("status"),
                "genres": show.get("genres") or [],
            }
        )
    return shows


def _normalize_episode(raw: dict, fallback_runtime: int) -> dict:
    return {
        "id": raw.get("id"),
        "season": raw.get("season"),
        "number": raw.get("number"),
        "title": raw.get("name") or f"Episode {raw.get('number')}",
        "runtime": raw.get("runtime") or fallback_runtime,
        "airdate": raw.get("airdate") or "",
        "summary": strip_html(raw.get("summary")),
        "type": raw.get("type") or "regular",
    }


def get_show(query: str) -> dict:
    """Fetch the best-matching show with its full episode list, normalized.

    Returns {"id", "name", "premiered", "status", "genres", "episodes": [...]}
    where episodes are ordered by (season, number) and specials without an
    episode number are dropped.
    """
    data = _get("/singlesearch/shows", {"q": query, "embed": "episodes"})
    fallback_runtime = data.get("averageRuntime") or DEFAULT_RUNTIME
    raw_eps = data.get("_embedded", {}).get("episodes", [])
    episodes = [
        _normalize_episode(e, fallback_runtime)
        for e in raw_eps
        if e.get("season") is not None and e.get("number") is not None
    ]
    episodes.sort(key=lambda e: (e["season"], e["number"]))
    return {
        "id": data.get("id"),
        "name": data.get("name"),
        "premiered": (data.get("premiered") or "")[:4],
        "status": data.get("status"),
        "genres": data.get("genres") or [],
        "episodes": episodes,
    }
